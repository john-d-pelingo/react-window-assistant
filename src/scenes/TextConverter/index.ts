export { TextConverter as default } from './TextConverter'
