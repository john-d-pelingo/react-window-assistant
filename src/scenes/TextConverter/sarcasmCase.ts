import { lowerCase, upperCase } from 'change-case'

const isLetter = (letter: string) => upperCase(letter) !== lowerCase(letter)

const LowerCase = 0 as const
const UpperCase = 1 as const

type Cases = typeof LowerCase | typeof UpperCase

export const sarcasmCase = (text: string, firstLetter: Cases = LowerCase) => {
  let textCase: Cases = firstLetter

  return text
    .split('')
    .map(letter => {
      if (letter.trim().length === 0 || !isLetter(letter)) {
        return letter
      }

      if (textCase === LowerCase) {
        textCase = UpperCase

        return lowerCase(letter)
      }

      textCase = LowerCase
      return upperCase(letter)
    })
    .join('')
}
