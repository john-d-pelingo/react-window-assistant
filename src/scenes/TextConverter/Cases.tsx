import { Chip } from '@material-ui/core'
import changeCase from 'change-case'
import React, { FC } from 'react'
import styled from 'styled-components'

import { sarcasmCase } from './sarcasmCase'
import {
  camel,
  Case,
  constant,
  dot,
  header,
  lower,
  lowerFirst,
  no,
  param,
  pascal,
  path,
  sarcasmLowerFirst,
  sarcasmUpperFirst,
  sentence,
  snake,
  swap,
  title,
  upper,
  upperFirst,
} from './useTextConverter/actions'

const Wrapper = styled.div`
  width: 80%;
  text-align: center;
  margin: 0 auto;
  margin-top: 24px;
`

const StyledChip = styled(Chip)`
  margin: 0 5px 5px;
`

const textCases: Array<[Case, string]> = [
  [camel, changeCase.camel(camel)],
  [constant, changeCase.constant(constant)],
  [dot, changeCase.dot(dot)],
  [header, changeCase.header(header)],
  [lower, changeCase.lower(lower)],
  [lowerFirst, changeCase.lcFirst(lowerFirst)],
  [no, changeCase.no(no)],
  [param, changeCase.param(param)],
  [pascal, changeCase.pascal(pascal)],
  [path, changeCase.path(path)],
  [sarcasmLowerFirst, sarcasmCase(sarcasmLowerFirst, 0)],
  [sarcasmUpperFirst, sarcasmCase(sarcasmUpperFirst, 1)],
  [sentence, changeCase.sentence(sentence)],
  [snake, changeCase.snake(snake)],
  [swap, changeCase.swap(swap)],
  [title, changeCase.title(title)],
  [upper, changeCase.upper(upper)],
  [upperFirst, changeCase.ucFirst(upperFirst)],
]

interface CasesProps {
  disabled?: boolean
  onSetCase: (textCase: Case) => void
}

export const Cases: FC<CasesProps> = ({ disabled = false, onSetCase }) => {
  return (
    <Wrapper>
      {textCases.map(([textCase, text]) => (
        <StyledChip
          key={textCase}
          disabled={disabled}
          label={text}
          onClick={() => {
            onSetCase(textCase)
          }}
          variant="outlined"
        />
      ))}
    </Wrapper>
  )
}
