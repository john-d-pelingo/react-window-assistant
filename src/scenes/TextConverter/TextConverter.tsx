import { Container, OutlinedInput } from '@material-ui/core'
import React, { ChangeEvent, FC, useEffect, useRef } from 'react'
import { Helmet } from 'react-helmet'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'

import { textConverterQueryParameter } from 'consants/strings'
import { extractQueryParameter } from 'helpers/extractQueryParameter'
import { useQueryParameterComputation } from 'hooks/useQueryParameterComputation'

import { Cases } from './Cases'
import { setCase, setText } from './useTextConverter/actions'
import { useTextConverter } from './useTextConverter/hook'

const StyledOutlinedInput = styled(OutlinedInput)`
  font-family: 'Courier New', mono, sans-serif;
  font-size: 16px;
`

// TODO: add tests
export const TextConverter: FC = () => {
  const inputNode = useRef<HTMLTextAreaElement>()
  const { dispatch, text } = useTextConverter({
    text: '',
  })

  const history = useHistory()
  useQueryParameterComputation({
    history,
    key: textConverterQueryParameter,
    value: text,
  })

  useEffect(() => {
    if (inputNode.current) {
      inputNode.current.focus()
    }

    const queryParameter = extractQueryParameter({
      history,
      key: textConverterQueryParameter,
    })

    if (queryParameter) {
      dispatch(setText(queryParameter))
    }
  }, [dispatch, history])

  const handleInputChange = (event: ChangeEvent<HTMLTextAreaElement>): void => {
    dispatch(setText(event.target.value))
  }

  return (
    <>
      <Helmet>
        <title>Text Converter</title>
      </Helmet>
      <Container maxWidth="xl" style={{ margin: '100px 0' }}>
        <StyledOutlinedInput
          fullWidth
          inputProps={{
            'aria-label': 'JSON input',
          }}
          inputRef={inputNode}
          labelWidth={0}
          multiline
          onChange={handleInputChange}
          placeholder="Write your text here"
          rows={20}
          value={text}
        />
        <Cases
          disabled={text.trim().length === 0}
          onSetCase={textCase => {
            dispatch(setCase(text, textCase))
          }}
        />
        {/* TODO: add clear and copy text */}
      </Container>
    </>
  )
}
