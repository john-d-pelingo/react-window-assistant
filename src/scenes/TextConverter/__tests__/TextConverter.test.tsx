import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { fireEvent, render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Helmet } from 'react-helmet'
import { Router } from 'react-router'

import { TextConverter } from '../TextConverter'

let memoryHistory = createMemoryHistory()

describe('scenes - TextConverter', () => {
  beforeEach(() => {
    memoryHistory = createMemoryHistory()
  })

  it('mounts', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { container } = render(
      <Router history={memoryHistory}>
        <TextConverter />
      </Router>,
    )

    expect(Helmet.peek().title).toBe('Text Converter')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('converts a written text to lower first, pascal, sarcasm lower first, upper and finally path case', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText, getByText } = render(
      <Router history={memoryHistory}>
        <TextConverter />
      </Router>,
    )

    const buttonLabels: string[] = [
      'camelCase',
      'CONSTANT_CASE',
      'dot.case',
      'Header-Case',
      'lower case',
      'lowerFirst case',
      'no case',
      'param-case',
      'PascalCase',
      'path/case',
      'sArCaSmLoWeRfIrSt CaSe',
      'SaRcAsMuPpErFiRsT cAsE',
      'Sentence case',
      'snake_case',
      'SWAP CASE',
      'Title Case',
      'UPPER CASE',
      'UpperFirst case',
    ]

    const textInputElement = getByLabelText('Text input') as HTMLTextAreaElement

    expect(document.activeElement).toEqual(textInputElement)

    buttonLabels.forEach((buttonLabel) => {
      const textCaseButtonElement = getByText(buttonLabel) as HTMLDivElement
      expect(textCaseButtonElement.parentNode).toHaveClass('Mui-disabled')
    })

    fireEvent.change(textInputElement, {
      target: {
        value:
          'Doggo ipsum you are doing me the shock floofs fluffer what a nice floof, snoot borkdrive he made many woofs pupper, doggo blep. Woofer extremely cuuuuuute woofer such treat vvv long water shoob floofs, shoober dat tungg tho lotsa pats blep.',
      },
    })

    expect(textInputElement).toHaveTextContent(
      'Doggo ipsum you are doing me the shock floofs fluffer what a nice floof, snoot borkdrive he made many woofs pupper, doggo blep. Woofer extremely cuuuuuute woofer such treat vvv long water shoob floofs, shoober dat tungg tho lotsa pats blep.',
    )
    expect(memoryHistory.location.search).toBe(
      '?text=Doggo%20ipsum%20you%20are%20doing%20me%20the%20shock%20floofs%20fluffer%20what%20a%20nice%20floof%2C%20snoot%20borkdrive%20he%20made%20many%20woofs%20pupper%2C%20doggo%20blep.%20Woofer%20extremely%20cuuuuuute%20woofer%20such%20treat%20vvv%20long%20water%20shoob%20floofs%2C%20shoober%20dat%20tungg%20tho%20lotsa%20pats%20blep.',
    )

    const lowerFirstCaseButtonElement = getByText(
      'lowerFirst case',
    ) as HTMLDivElement
    fireEvent.click(lowerFirstCaseButtonElement)

    expect(textInputElement).toHaveTextContent(
      'doggo ipsum you are doing me the shock floofs fluffer what a nice floof, snoot borkdrive he made many woofs pupper, doggo blep. Woofer extremely cuuuuuute woofer such treat vvv long water shoob floofs, shoober dat tungg tho lotsa pats blep.',
    )
    expect(memoryHistory.location.search).toBe(
      '?text=doggo%20ipsum%20you%20are%20doing%20me%20the%20shock%20floofs%20fluffer%20what%20a%20nice%20floof%2C%20snoot%20borkdrive%20he%20made%20many%20woofs%20pupper%2C%20doggo%20blep.%20Woofer%20extremely%20cuuuuuute%20woofer%20such%20treat%20vvv%20long%20water%20shoob%20floofs%2C%20shoober%20dat%20tungg%20tho%20lotsa%20pats%20blep.',
    )

    const sarcasmLowerFirstCaseButtonElement = getByText(
      'sArCaSmLoWeRfIrSt CaSe',
    ) as HTMLDivElement
    fireEvent.click(sarcasmLowerFirstCaseButtonElement)

    expect(textInputElement).toHaveTextContent(
      'dOgGo IpSuM yOu ArE dOiNg Me ThE sHoCk FlOoFs FlUfFeR wHaT a NiCe FlOoF, sNoOt BoRkDrIvE hE mAdE mAnY wOoFs PuPpEr, DoGgO bLeP. wOoFeR eXtReMeLy CuUuUuUtE wOoFeR sUcH tReAt VvV lOnG wAtEr ShOoB fLoOfS, sHoObEr DaT tUnGg ThO lOtSa PaTs BlEp.',
    )
    expect(memoryHistory.location.search).toBe(
      '?text=dOgGo%20IpSuM%20yOu%20ArE%20dOiNg%20Me%20ThE%20sHoCk%20FlOoFs%20FlUfFeR%20wHaT%20a%20NiCe%20FlOoF%2C%20sNoOt%20BoRkDrIvE%20hE%20mAdE%20mAnY%20wOoFs%20PuPpEr%2C%20DoGgO%20bLeP.%20wOoFeR%20eXtReMeLy%20CuUuUuUtE%20wOoFeR%20sUcH%20tReAt%20VvV%20lOnG%20wAtEr%20ShOoB%20fLoOfS%2C%20sHoObEr%20DaT%20tUnGg%20ThO%20lOtSa%20PaTs%20BlEp.',
    )

    const upperCaseButtonElement = getByText('UPPER CASE') as HTMLDivElement
    fireEvent.click(upperCaseButtonElement)

    expect(textInputElement).toHaveTextContent(
      'DOGGO IPSUM YOU ARE DOING ME THE SHOCK FLOOFS FLUFFER WHAT A NICE FLOOF, SNOOT BORKDRIVE HE MADE MANY WOOFS PUPPER, DOGGO BLEP. WOOFER EXTREMELY CUUUUUUTE WOOFER SUCH TREAT VVV LONG WATER SHOOB FLOOFS, SHOOBER DAT TUNGG THO LOTSA PATS BLEP.',
    )
    expect(memoryHistory.location.search).toBe(
      '?text=DOGGO%20IPSUM%20YOU%20ARE%20DOING%20ME%20THE%20SHOCK%20FLOOFS%20FLUFFER%20WHAT%20A%20NICE%20FLOOF%2C%20SNOOT%20BORKDRIVE%20HE%20MADE%20MANY%20WOOFS%20PUPPER%2C%20DOGGO%20BLEP.%20WOOFER%20EXTREMELY%20CUUUUUUTE%20WOOFER%20SUCH%20TREAT%20VVV%20LONG%20WATER%20SHOOB%20FLOOFS%2C%20SHOOBER%20DAT%20TUNGG%20THO%20LOTSA%20PATS%20BLEP.',
    )

    const pathCaseButtonElement = getByText('path/case') as HTMLDivElement
    fireEvent.click(pathCaseButtonElement)

    expect(textInputElement).toHaveTextContent(
      'doggo/ipsum/you/are/doing/me/the/shock/floofs/fluffer/what/a/nice/floof/snoot/borkdrive/he/made/many/woofs/pupper/doggo/blep/woofer/extremely/cuuuuuute/woofer/such/treat/vvv/long/water/shoob/floofs/shoober/dat/tungg/tho/lotsa/pats/blep',
    )
    expect(memoryHistory.location.search).toBe(
      '?text=doggo%2Fipsum%2Fyou%2Fare%2Fdoing%2Fme%2Fthe%2Fshock%2Ffloofs%2Ffluffer%2Fwhat%2Fa%2Fnice%2Ffloof%2Fsnoot%2Fborkdrive%2Fhe%2Fmade%2Fmany%2Fwoofs%2Fpupper%2Fdoggo%2Fblep%2Fwoofer%2Fextremely%2Fcuuuuuute%2Fwoofer%2Fsuch%2Ftreat%2Fvvv%2Flong%2Fwater%2Fshoob%2Ffloofs%2Fshoober%2Fdat%2Ftungg%2Ftho%2Flotsa%2Fpats%2Fblep',
    )
  })

  it(`removes the text query parameter when the text input is empty/whitespace`, () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <TextConverter />
      </Router>,
    )

    const textInputElement = getByLabelText('Text input') as HTMLTextAreaElement

    fireEvent.change(textInputElement, {
      target: {
        value: 'Text, is it?',
      },
    })

    expect(textInputElement).toHaveTextContent('Text, is it?')
    expect(memoryHistory.location.search).toBe('?text=Text%2C%20is%20it%3F')

    fireEvent.change(textInputElement, {
      target: {
        value: ' \t\t\t\r',
      },
    })

    expect(textInputElement).toHaveTextContent('')
    expect(memoryHistory.location.search).toBe('')
  })
})
