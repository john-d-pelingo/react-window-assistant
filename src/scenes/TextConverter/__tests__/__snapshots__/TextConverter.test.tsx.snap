// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes - TextConverter mounts 1`] = `
.c1 {
  width: 80%;
  text-align: center;
  margin: 0 auto;
  margin-top: 24px;
}

.c2 {
  margin: 0 5px 5px;
}

.c0 {
  font-family: 'Courier New',mono,sans-serif;
  font-size: 16px;
}

<div
  class="MuiContainer-root MuiContainer-maxWidthXl"
  style="margin: 100px 0px;"
>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root c0 MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-multiline MuiOutlinedInput-multiline"
  >
    <textarea
      aria-label="Text input"
      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
      placeholder="Write your text here"
      rows="20"
    />
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-42 MuiOutlinedInput-notchedOutline"
      style="padding-left: 8px;"
    >
      <legend
        class="PrivateNotchedOutline-legend-43"
        style="width: 0.01px;"
      >
        <span>
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
  <div
    class="c1"
  >
    <div
      aria-disabled="true"
      aria-label="camelCase"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        camelCase
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="CONSTANT_CASE"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        CONSTANT_CASE
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="dot.case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        dot.case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="Header-Case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        Header-Case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="lower case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        lower case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="lowerFirst case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        lowerFirst case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="no case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        no case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="param-case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        param-case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="PascalCase"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        PascalCase
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="path/case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        path/case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="sArCaSmLoWeRfIrSt CaSe"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        sArCaSmLoWeRfIrSt CaSe
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="SaRcAsMuPpErFiRsT cAsE"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        SaRcAsMuPpErFiRsT cAsE
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="Sentence case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        Sentence case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="snake_case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        snake_case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="SWAP CASE"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        SWAP CASE
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="Title Case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        Title Case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="UPPER CASE"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        UPPER CASE
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="true"
      aria-label="UpperFirst case"
      class="MuiButtonBase-root MuiChip-root c2 MuiChip-outlined Mui-disabled MuiChip-clickable"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiChip-label"
      >
        UpperFirst case
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </div>
</div>
`;
